### ISSUE TITLE
monotonic_cst is not strictly satisfied 

### ISSUE THREAD

**Issue by yzhang3000** [zillow/quantile-forest#73]
I tried to add monotonic_cst to the RandomForestQuantileRegressor(), however, the prediction shows violation of the monotonicity requirement in some cases. 

Comparing the models with and without monotonic_cst, the monotonic_cst argument does have effect on the models, however, the monotonic requirements are not always satisfied with testing data.

As the comparison, the model trained using sklearn.ensemble.RandomForestRegressor() guarantees the monotonic requirements with the same data set.

**Comment by reidjohnson** [comment]
Thanks for the issue. Would you happen to be able to provide the code for a minimally reproducible example?

**Comment by yzhang3000** [comment]
# RandomForestQuantileRegressor monotonic constraints testing
---


```python
from quantile_forest import RandomForestQuantileRegressor
from sklearn.ensemble import RandomForestRegressor
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
```

## Generate testing data


```python
N = 2000
xs = np.random.rand(N, 2) 
ys = xs[:,0] + np.random.normal(size=N) * 0.2
X_train = pd.DataFrame(data=xs, columns=['a', 'b'])
```

## 1. RandomForestQuantileRegressor without monotonic constraints


```python
rfqmodel = RandomForestQuantileRegressor(
    monotonic_cst = [0, 0],
        )

rfqmodel.fit(X_train, ys)

fig, ax = plt.subplots()
ax.scatter(X_train.a, ys, s=5, alpha=0.7)
n = 200
xt = np.array([np.linspace(0, 1, n), np.ones(n)*0.5]).T
xt = pd.DataFrame(xt, columns=['a', 'b'])
y_pred = rfqmodel.predict(xt)
ax.plot(xt.a, y_pred, 'r',)
ax.grid()
```

    
![output_5_1](https://github.com/user-attachments/assets/4b82c87a-aece-48e1-8625-5eefb385554d)    


## 2. RandomForestQuantileRegressor with monotonic constraints


```python
rfqmodel = RandomForestQuantileRegressor(
    monotonic_cst = [1, 0],
        )

rfqmodel.fit(X_train, ys)

fig, ax = plt.subplots()
ax.scatter(X_train.a, ys, s=5, alpha=0.7)
n = 200
xt = np.array([np.linspace(0, 1, n), np.ones(n)*0.5]).T
xt = pd.DataFrame(xt, columns=['a', 'b'])
y_pred = rfqmodel.predict(xt)
ax.plot(xt.a, y_pred, 'r', )
ax.grid()
```
    
![output_7_1](https://github.com/user-attachments/assets/144a2de2-9af6-4c3a-b988-42c62920d4f9)


## 3. RandomForestRegressor with monotonic constraints


```python
rfmodel = RandomForestRegressor(
    monotonic_cst = [1, 0],
        )

rfmodel.fit(X_train, ys)

fig, ax = plt.subplots()
ax.scatter(X_train.a, ys, s=5, alpha=0.7)

n = 200
xt = np.array([np.linspace(0, 1, n), np.ones(n)*0.5]).T
xt = pd.DataFrame(xt, columns=['a', 'b'])
y_pred = rfmodel.predict(xt)
ax.plot(xt.a, y_pred, 'r', )
ax.grid()
```



![output_9_0](https://github.com/user-attachments/assets/1e79e4a6-a89d-4011-bc8b-2142edec7171)    



```python

```

**Comment by reidjohnson** [comment]
@yzhang3000 Thank you, will look into this!

**Comment by reidjohnson** [comment]
@yzhang3000 Thanks again for reporting this issue. It should now be fixed in the latest release, version `1.3.8`.
